pkg/controller/atlasbackupcompliancepolicy/atlasbackupcompliancepolicy_controller.go

```diff
@@ -0,0 +1,177 @@
+package atlasbackupcompliancepolicy
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api"
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/atlas"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/customresource"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/statushandler"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/workflow"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/indexer"
+)
+
+type AtlasBackupCompliancePolicyReconciler struct {
+	Client                      client.Client
+	Scheme                      *runtime.Scheme
+	GlobalPredicates            []predicate.Predicate
+	EventRecorder               record.EventRecorder
+	AtlasProvider               atlas.Provider
+	Log                         *zap.SugaredLogger
+	ObjectDeletionProtection    bool
+	SubObjectDeletionProtection bool
+}
+
+func (r *AtlasBackupCompliancePolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.With("atlasbackupcompliancepolicy", req.NamespacedName)
+	log.Infow("-> Starting AtlasBackupCompliancePolicy reonciliation")
+
+	bcp := akov2.AtlasBackupCompliancePolicy{}
+	result := customresource.PrepareResource(ctx, r.Client, req, &bcp, log)
+	if result.IsOk() {
+		return result.ReconcileResult(), nil
+	}
+
+	return r.ensureAtlasBackupCompliancePolicy(ctx, log, &bcp)
+}
+
+func (r *AtlasBackupCompliancePolicyReconciler) ensureAtlasBackupCompliancePolicy(ctx context.Context, log *zap.SugaredLogger, bcp *akov2.AtlasBackupCompliancePolicy) (ctrl.Result, error) {
+	if customresource.ReconciliationShouldBeSkipped(bcp) {
+		return r.skip(ctx, log, bcp), nil
+	}
+
+	conditions := akov2.InitCondition(bcp, api.FalseCondition(api.ReadyType))
+	workflowCtx := workflow.NewContext(log, conditions, ctx)
+	defer statushandler.Update(workflowCtx, r.Client, r.EventRecorder, bcp)
+
+	isValid := customresource.ValidateResourceVersion(workflowCtx, bcp, r.Log)
+	if !isValid.IsOk() {
+		return r.invalidate(isValid)
+	}
+
+	if !r.AtlasProvider.IsResourceSupported(bcp) {
+		return r.unsupport(workflowCtx)
+	}
```

Comment:
Nitpick

I wonder if all these checks are just a common preamble which is not part of the main resource state machine designed.

I would probably strip them away back to the main reconcile method, something like this:

```suggestion
	}

	if customresource.ReconciliationShouldBeSkipped(bcp) {
		return r.skip(ctx, log, bcp), nil
	}

	conditions := akov2.InitCondition(bcp, api.FalseCondition(api.ReadyType))
	workflowCtx := workflow.NewContext(log, conditions, ctx)
	defer statushandler.Update(workflowCtx, r.Client, r.EventRecorder, bcp)

	isValid := customresource.ValidateResourceVersion(workflowCtx, bcp, r.Log)
	if !isValid.IsOk() {
		return r.invalidate(isValid)
	}

	if !r.AtlasProvider.IsResourceSupported(bcp) {
		return r.unsupport(workflowCtx)
	}

	return r.ensureAtlasBackupCompliancePolicy(ctx, log, &bcp)
}

func (r *AtlasBackupCompliancePolicyReconciler) ensureAtlasBackupCompliancePolicy(ctx context.Context, log *zap.SugaredLogger, bcp *akov2.AtlasBackupCompliancePolicy) (ctrl.Result, error) {
```

Reply:
very good suggestion 👍